### PR TITLE
Add incident/retrospective/remediation models and migrations

### PR DESCRIPTION
**Pull request by joshuatobin**
This is mostly exploratory for what models we should stub out. 

CC: @heroku/sre

### CODE REVIEW

File: app/models/incident.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class Incident < ActiveRecord::Base
+  has_many :retrospectives
+  has_many :remediations, through: :retrospectives
+
+  validates :title, presence: true
```
Comment (lexelby) [inline app/models/incident.rb]: This is going to sound a bit weird, but I think we shouldn't validate the presence of the title.  Here's the scenario I'm thinking of:

1. Something fishy starts going on, an engineer pages IC.
2. IC does `h start incident`.
3. Retrodot sees that no incident is open right now so assumes the IC is declaring the start of a new incident that hasn't been posted to the status site yet.
4. Retrodot starts a new timeline for a new incident.
5. IC has support post the incident to the status site.

In other words, it may be necessary to create an incident object without a title.

File: db/migrate/20160805132331_create_incidents.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+class CreateIncidents < ActiveRecord::Migration[5.0]
+  def change
+    create_table :incidents do |t|
+      t.date       :date
+      t.string     :title
+      t.integer    :issue
```
Comment (lexelby) [inline db/migrate/20160805132331_create_incidents.rb]: Can we set this as the primary key?

File: app/models/incident.rb (hunk 1 above)
Comment (joshuatobin) [inline app/models/incident.rb]: I guess the sync would just then eventually fill in the title?

File: db/migrate/20160805132331_create_incidents.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+      t.timestamps null: false
+    end
+
+    create_table :remediations do |t|
+      t.date       :date
+      t.string     :description
```
Comment (lexelby) [inline db/migrate/20160805132331_create_incidents.rb]: can we add a text-search index on this?

File: db/migrate/20160805132331_create_incidents.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      t.boolean    :completed
+      t.string     :pull_request
+      t.timestamps null: false
+
+      t.belongs_to :retrospective
+    end
+
+    create_table :retrospectives do |t|
+      t.date       :date
+      t.string     :description
```
Comment (lexelby) [inline db/migrate/20160805132331_create_incidents.rb]: text-search here too

File: app/models/incident.rb (hunk 1 above)
Comment (lexelby) [inline app/models/incident.rb]: Yup.

**Comment by lexelby** [comment]
Let's go ahead and add fields that are currently in [Past Incidents](https://docs.google.com/spreadsheets/d/1d4wYxLD42ickxauV9NJEdPmAKkijIk9Gc1JjT5kYKdM/edit#gid=0):

* date of retro
* category (probably join to a categories table)
* primary team (join to a teams table, fill in with the teams we have now, or use DW data if we can)
* needs public followup
* date of public followup
* primary component (use DW data, or we can just leave this blank for now, maybe?)

At the very least, we should have date of retro, category, and public followup stuff to achieve parity with Past Incidents.

File: app/models/incident.rb (hunk 1 above)
Comment (joshuatobin) [inline app/models/incident.rb]: Fixed in 3278bc1

File: db/migrate/20160805132331_create_incidents.rb (hunk 2 above)
Comment (joshuatobin) [inline db/migrate/20160805132331_create_incidents.rb]: @reidmix Do you have thoughts on this. Everything I'm reading suggests /not/ doing this. Is it a common rails practice?

File: db/migrate/20160805132331_create_incidents.rb (hunk 3 above)
Comment (joshuatobin) [inline db/migrate/20160805132331_create_incidents.rb]: @lexelby I've not done this before. Got an example?

File: db/migrate/20160805132331_create_incidents.rb (hunk 2 above)
Comment (jmervine) [inline db/migrate/20160805132331_create_incidents.rb]: Yes and no. I would remove this field all together and use the default primary key of `id` as the issue id.

Comment (joshuatobin) [inline db/migrate/20160805132331_create_incidents.rb]: @jmervine Either way we need that field to record the issue # from status.heroku.com

Comment (jmervine) [inline db/migrate/20160805132331_create_incidents.rb]: Ah, I see. Then yeah, it shouldn't be the primary key, although indexed for sure.

File: app/models/retrospective.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class Retrospective < ActiveRecord::Base
+  has_many   :remediations
+  belongs_to :incident
+
+  validates :incident_id, :date, :description, presence: true
```
Comment (reidmix) [inline app/models/retrospective.rb]: You can validate on `:incident` which usually makes things easier down the line.

File: db/migrate/20160805132331_create_incidents.rb (hunk 2 above)
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: So, typically (not always), I let the table have it's own (artificial) PK.  Then I reference the external number (I might even call it `:status_id` or `:status_incident_id` to be super explicit).  

The reason I wouldn't use the external ID it as the PK is it paints you into a corner, esp. when rails gets super finicky about IDs, but also because then you get very beholden to another external system's way of numbering things as your way of numbering things.

File: db/migrate/20160805132331_create_incidents.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+class CreateIncidents < ActiveRecord::Migration[5.0]
+  def change
+    create_table :incidents do |t|
+      t.date       :date
```
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: Is this the starting date of the incident?  Can naming better reflect it?

File: db/migrate/20160805132331_create_incidents.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
+class CreateIncidents < ActiveRecord::Migration[5.0]
```
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: I might be inclined to break these out in separate migrations -- but that's me -- maybe have the title better reflect the migration?

File: db/migrate/20160805132331_create_incidents.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  def change
+    create_table :incidents do |t|
```
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: I think you need `duration` as well -- are there other fields?

File: db/migrate/20160805132331_create_incidents.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      t.date       :date
+      t.string     :title
+      t.integer    :issue
+
+      t.timestamps null: false
+    end
+
+    create_table :remediations do |t|
```
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: Do you need remediations at this time?
I'd say boot it to keep things simple for now.

File: db/migrate/20160805132331_create_incidents.rb (hunk 3 above)
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: @lexelby what's text-search? please to teach :D

File: db/migrate/20160805132331_create_incidents.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+      t.belongs_to :retrospective
+    end
+
+    create_table :retrospectives do |t|
+      t.date       :date
+      t.string     :description
+      t.timestamps null: false
+
+      t.belongs_to :incident
+      t.belongs_to :remediation
```
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: Nice, I always forget about these, very nice.

**Comment by reidmix** [comment]
Added my feedback, great start!

File: db/migrate/20160805132331_create_incidents.rb (hunk 7 above)
Comment (joshuatobin) [inline db/migrate/20160805132331_create_incidents.rb]: @reidmix 👍 Done. I created 4 separate migrations.

File: db/migrate/20160805132331_create_incidents.rb (hunk 8 above)
Comment (joshuatobin) [inline db/migrate/20160805132331_create_incidents.rb]: Fixed as well.

**Comment by joshuatobin** [comment]
@lexelby I've added a categories table and updated incidents to reflect the missing columns. I'm going to deal with teams and components separately in another PR. I need to get this PR in as its blocking on the syncher.

File: db/migrate/20160805132331_create_incidents.rb [hunk 11]
```diff
@@ -0,0 +1,15 @@
+class CreateIncidents < ActiveRecord::Migration[5.0]
+  def change
+    create_table :incidents do |t|
+      t.date       :created
+      t.date       :resolved
```
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: Are these a date or a date/time (or does date include time here -- i forget).

Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: You can use the rails naming conventions here too, but probably need to rename `:created` because of timestamps naming: `:started_on`, `:resolved_on`

File: db/migrate/20160805132331_create_incidents.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+      t.integer    :duration
+      t.string     :state
+      t.string     :title
+      t.integer    :issue
```
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: I'd probably call this `:status_id` or `:status_incident_id` (but at least add "_id").

File: db/migrate/20160805132331_create_incidents.rb [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+      t.boolean    :public_followup
+      t.date       :public_followup_date
```
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: Here's a good example of rails naming conventions with a date, I'd say `:public_followup_on`

Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: Personally, when naming, I don't usually like putting the type in the name of the variable/column_name/etc. 
Mostly because its duplicative, but I don't think it adds anything necessarily.

File: db/migrate/20160809132715_create_remediations.rb [hunk 14]
```diff
@@ -0,0 +1,13 @@
+class CreateRemediations < ActiveRecord::Migration[5.0]
+  def change
+    create_table :remediations do |t|
+      t.date       :date
```
Comment (reidmix) [inline db/migrate/20160809132715_create_remediations.rb]: What is it a date of?  Is this the created date, the due date, etc.? 😄

File: db/migrate/20160809132715_create_remediations.rb [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+      t.string     :description
+      t.boolean    :completed
+      t.string     :pull_request
```
Comment (reidmix) [inline db/migrate/20160809132715_create_remediations.rb]: Oh interesting.  I like this.

File: db/migrate/20160809132722_create_retrospectives.rb [hunk 16]
```diff
@@ -0,0 +1,12 @@
+class CreateRetrospectives < ActiveRecord::Migration[5.0]
+  def change
+    create_table :retrospectives do |t|
+      t.date       :date
+      t.string     :description
+      t.timestamps null: false
+
+      t.belongs_to :incident
+      t.belongs_to :remediation
```
Comment (reidmix) [inline db/migrate/20160809132722_create_retrospectives.rb]: Don't need the `remediation` association -- take a look at the `schema.rb`

Comment (reidmix) [inline db/migrate/20160809132722_create_retrospectives.rb]: The rest of the associations look good to me.

File: app/models/incident.rb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+class Incident < ActiveRecord::Base
+  has_many :retrospectives
+  has_many :remediations, through: :retrospectives
+  has_one  :categories
```
Comment (reidmix) [inline app/models/incident.rb]: I think you're looking for:
```ruby
has_one :category
```

Rails should do the right thing.

File: app/models/category.rb [hunk 18]
```diff
@@ -0,0 +1,4 @@
+class Category < ActiveRecord::Base
+  has_many :incidents
+  belongs_to :incidents
+end
```
Comment (reidmix) [inline app/models/category.rb]: I believe you just need:
```ruby
belongs_to :incident
```
Also, singular

"`belongs_to` will place the foreign key in the declaring model whereas `has_one` will place it in the other model."

File: app/models/incident.rb (hunk 17 above)
Comment (reidmix) [inline app/models/incident.rb]: If the usecase is that an incident can only have a single category I think you want:
```ruby
belongs_to :category
```
It's also singular, rails will do the right thing.

File: app/models/category.rb [hunk 19]
```diff
@@ -0,0 +1,4 @@
+class Category < ActiveRecord::Base
+  has_many :incidents
+  belongs_to :incidents
```
Comment (reidmix) [inline app/models/category.rb]: If the usecase is that an incident can only have a single category, you'll just need:
```ruby
has_many :incident
```

File: db/migrate/20160809134428_create_categories.rb [hunk 20]
```diff
@@ -0,0 +1,9 @@
+class CreateCategories < ActiveRecord::Migration[5.0]
+  def change
+    create_table :categories do |t|
+      t.string   :name
+
+      t.belongs_to :incident
```
Comment (reidmix) [inline db/migrate/20160809134428_create_categories.rb]: If the usecase is that an incident can only have a single category, this'll need to be a `has_many`
which then matches the model association too (I know you were working through some ambiguity there)

File: db/migrate/20160809132722_create_retrospectives.rb (hunk 16 above)
Comment (reidmix) [inline db/migrate/20160809132722_create_retrospectives.rb]: You probably need to add (move) the category association here:
```ruby
t.belongs_to :category
```
And remove it from the "categories" migration

File: db/migrate/20160809134428_create_categories.rb (hunk 20 above)
Comment (reidmix) [inline db/migrate/20160809134428_create_categories.rb]: Remove (move) this, and put it in the "retrospectives" migration.

File: db/migrate/20160809132722_create_retrospectives.rb (hunk 16 above)
Comment (reidmix) [inline db/migrate/20160809132722_create_retrospectives.rb]: You want to do this because you want the FK to be in this table, so that each Incident has only one Category, and that any Category can be set on an Incident.

**Comment by reidmix** [comment]
I noticed a few more things this round, which was mostly around naming and associations.

Nice to see these tables getting fleshed out, it's exciting to see.

File: app/models/incident.rb (hunk 17 above)
Comment (joshuatobin) [inline app/models/incident.rb]: Ugh I always get these messed up. Fixed in 72277ae

File: db/migrate/20160805132331_create_incidents.rb (hunk 13 above)
Comment (joshuatobin) [inline db/migrate/20160805132331_create_incidents.rb]: Noted. 👍

File: db/migrate/20160809132715_create_remediations.rb (hunk 14 above)
Comment (joshuatobin) [inline db/migrate/20160809132715_create_remediations.rb]: Fixed in 81269dd

File: db/migrate/20160805132331_create_incidents.rb (hunk 13 above)
Comment (joshuatobin) [inline db/migrate/20160805132331_create_incidents.rb]: Fixed in 81269dd

File: db/migrate/20160805132331_create_incidents.rb (hunk 12 above)
Comment (joshuatobin) [inline db/migrate/20160805132331_create_incidents.rb]: Fixed in 81269dd

File: db/migrate/20160805132331_create_incidents.rb (hunk 11 above)
Comment (joshuatobin) [inline db/migrate/20160805132331_create_incidents.rb]: Fixed in 81269dd

File: db/migrate/20160809132722_create_retrospectives.rb (hunk 16 above)
Comment (joshuatobin) [inline db/migrate/20160809132722_create_retrospectives.rb]: @reidmix Would I need to remove the `belongs_to :category` in the incident model as well?

Comment (joshuatobin) [inline db/migrate/20160809132722_create_retrospectives.rb]: Fixed in bff803d

**Comment by joshuatobin** [comment]
@reidmix I think I've addressed all your suggestions

File: app/models/remediations.rb [hunk 21]
```diff
@@ -0,0 +1,5 @@
+class Remediation < ActiveRecord::Base
+  belongs_to :incident
```
Comment (lexelby) [inline app/models/remediations.rb]: This is probably my lack of activerecord knowledge talking, but why do we need the `belongs_to :incident` when we have `has_one :incident, through: :retrospective`?

File: db/migrate/20160805132331_create_incidents.rb [hunk 22]
```diff
@@ -0,0 +1,15 @@
+class CreateIncidents < ActiveRecord::Migration[5.0]
+  def change
+    create_table :incidents do |t|
+      t.datetime   :started_at
+      t.datetime   :resolved_at
+      t.integer    :duration
+      t.string     :state
+      t.string     :title
+      t.integer    :status_incident_id
```
Comment (lexelby) [inline db/migrate/20160805132331_create_incidents.rb]: Hmm, are you thinking that this won't be the same as the primary key ID in retrodot?  I'm concerned that having two separate IDs for an incident may lead to confusion and coding errors.  IMO having `incident.id` in Retrodot match `incident.id` in Status is a good idea.

File: db/migrate/20160809132715_create_remediations.rb [hunk 23]
```diff
@@ -0,0 +1,13 @@
+class CreateRemediations < ActiveRecord::Migration[5.0]
+  def change
+    create_table :remediations do |t|
+      t.date       :created_on
+      t.string     :description
+      t.boolean    :completed
+      t.string     :pull_request
```
Comment (lexelby) [inline db/migrate/20160809132715_create_remediations.rb]: I'm not sure about this.  I would think big remediations could easily require multiple pull requests?

Comment (joshuatobin) [inline db/migrate/20160809132715_create_remediations.rb]: We can change this later if thats the case.

File: db/migrate/20160805132331_create_incidents.rb (hunk 22 above)
Comment (joshuatobin) [inline db/migrate/20160805132331_create_incidents.rb]: See previous comments about this.

File: app/models/remediations.rb (hunk 21 above)
Comment (joshuatobin) [inline app/models/remediations.rb]: @lex yeah you're actually right. It ended up causing me grief in the console so I removed it. 


```[38] pry(main)> Remediation.create!(description: 'hihi', completed: true, retrospective_id: 2, created_on: DateTime.now)
   (0.1ms)  BEGIN
  Incident Load (0.4ms)  SELECT  "incidents".* FROM "incidents" INNER JOIN "retrospectives" ON "incidents"."id" = "retrospectives"."incident_id" WHERE "retrospectives"."id" = $1 LIMIT $
2  [["id", 2], ["LIMIT", 1]]
  Retrospective Load (0.2ms)  SELECT  "retrospectives".* FROM "retrospectives" WHERE "retrospectives"."id" = $1 LIMIT $2  [["id", 2], ["LIMIT", 1]]
   (0.1ms)  ROLLBACK
  Incident Load (0.3ms)  SELECT "incidents".* FROM "incidents" INNER JOIN "retrospectives" ON "incidents"."id" = "retrospectives"."incident_id" WHERE "retrospectives"."id" = $1  [["id",
 2]]
NoMethodError: undefined method `updated?' for #<ActiveRecord::Associations::HasOneThroughAssociation:0x007f92dc82bab0>
```

File: db/migrate/20160805132331_create_incidents.rb (hunk 3 above)
Comment (lexelby) [inline db/migrate/20160805132331_create_incidents.rb]: No idea!  I was hoping you folks knew. ;)  I think postgres might have a feature for doing full-text search, I know MySQL does.  If so, we want to use it here.  If not, we're looking at something like ElasticSearch (yikes).

**Comment by joshuatobin** [comment]
```
Incident.create!(state: 'resolved', title: 'api is busted', public_followup: false, started_at: '10/12/2015', status_incident_id: '992', duration: 8888)
#<Incident:0x007f92d44d1098
 id: 1,
 started_at: Thu, 10 Dec 2015 00:00:00 UTC +00:00,
 resolved_at: nil,
 duration: 8888,
 state: "resolved",
 title: "api is busted",
 status_incident_id: 992,
 public_followup: false,
 public_followup_on: nil,
 created_at: Tue, 09 Aug 2016 19:14:14 UTC +00:00,
 updated_at: Tue, 09 Aug 2016 19:14:14 UTC +00:00>

# Create Category
Category.create!(name: 'my category')
#<Category:0x007f92da8a15f0 id: 1, name: "my category">

#Create Retro
Retrospective.create!(created_on: DateTime.now, description: 'hihi', incident_id: 1, category_id: 1)
#<Retrospective:0x007f92d4e0e868
 id: 1,
 created_on: Tue, 09 Aug 2016 19:16:16 UTC +00:00,
 description: "hihi",
 created_at: Tue, 09 Aug 2016 19:16:16 UTC +00:00,
 updated_at: Tue, 09 Aug 2016 19:16:16 UTC +00:00,
 incident_id: 1,
 category_id: 1>

# Create Remedation
Remediation.create!(created_on: DateTime.now, description: 'hihihi', completed: true, retrospective_id: 1)
id: 1,
 created_on: Tue, 09 Aug 2016 19:16:43 UTC +00:00,
 description: "hihihi",
 completed: true,
 pull_request: nil,
 created_at: Tue, 09 Aug 2016 19:16:43 UTC +00:00,
 updated_at: Tue, 09 Aug 2016 19:16:43 UTC +00:00,
 retrospective_id: 1>

# Verify 
Incident.first.retrospectives
=> [#<Retrospective:0x007f92d474c890
  id: 1,
  created_on: Tue, 09 Aug 2016 19:16:16 UTC +00:00,
  description: "hihi",
  created_at: Tue, 09 Aug 2016 19:16:16 UTC +00:00,
  updated_at: Tue, 09 Aug 2016 19:16:16 UTC +00:00,
  incident_id: 1,
  category_id: 1>]

Incident.first.retrospectives.first.incident
=> #<Incident:0x007f92d64e7238
 id: 1,
 started_at: Thu, 10 Dec 2015 00:00:00 UTC +00:00,
 resolved_at: nil,
 duration: 8888,
...

 Incident.first.remediations
[#<Remediation:0x007f92d6522dd8
  id: 1,
  created_on: Tue, 09 Aug 2016 19:16:43 UTC +00:00,
  description: "hihihi",
  completed: true,
  pull_request: nil,
  created_at: Tue, 09 Aug 2016 19:16:43 UTC +00:00,
  updated_at: Tue, 09 Aug 2016 19:16:43 UTC +00:00,
  retrospective_id: 1>]

Category.first.retrospectives
=> [#<Retrospective:0x007f92dc890690
  id: 1,
  created_on: Tue, 09 Aug 2016 19:16:16 UTC +00:00,
  description: "hihi",
  created_at: Tue, 09 Aug 2016 19:16:16 UTC +00:00,
  updated_at: Tue, 09 Aug 2016 19:16:16 UTC +00:00,
  incident_id: 1,
  category_id: 1>]
```

File: db/migrate/20160809132715_create_remediations.rb (hunk 23 above)
Comment (lexelby) [inline db/migrate/20160809132715_create_remediations.rb]: Fine with me, just wanted to call it out.

**Comment by lexelby** [comment]
:+1:

File: app/models/remediations.rb (hunk 21 above)
Comment (reidmix) [inline app/models/remediations.rb]: Yup, 👍

File: db/migrate/20160805132331_create_incidents.rb (hunk 3 above)
Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: I think string is ok.  Not sure why text-search came up?

Comment (lexelby) [inline db/migrate/20160805132331_create_incidents.rb]: @reidmix I'm sure we're going to want to be able to search in incident and retrospective history.  It's something we've been lacking in the existing setup for a long time.  A standard `LIKE` query to search for single terms requires a full table scan, which may not be very fast.  I guess for the relatively small number of rows in the incidents table, though, it may be fine.

Comment (reidmix) [inline db/migrate/20160805132331_create_incidents.rb]: Ah i understand now..  Yeah so when we cross that bridge there are a few routes we can take, and we can use something like elasticsearch, I see where you were going with that now.  

Agreed, strings should be fine for now, we can change things up in the future when we need to.
